### ISSUE TITLE
Dropout layers break if set_learning_phase has been called

### ISSUE THREAD

**Issue by toddgardner** [fchollet/keras#5268]
If set_learning_phase has been called the next layer in the model gets an error, as described here:

http://stackoverflow.com/questions/41704617/exception-extending-keras-vgg16-model-self-assert-input-compatibilityx-throw

(included a full reproducer below)

I believe this is the result of the change: https://github.com/fchollet/keras/commit/8c0c3774e6cf88704f685784f8baba9694220d4d#diff-05ec8fc63f0e9bb35f99c0ca76a66d0e

In particular, if set_learning_phase has been called, call returns a function, not a tensor, because `in_train_phase` does not evaluate the callable if set_learn_phase is called.

The most robust fix is probably to make in_train/test_phase accept a callable, like switch does, and evaluate it in case set_learning_phase has been called. Not sure how to test it but can through together a PR.

- [*] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [*] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ ] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [*] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

```
import keras
from keras import backend as K
from keras.layers import Convolution2D, MaxPooling2D, Flatten, Input
from keras.layers import Dense, Dropout, Reshape, Input
import numpy as np
from keras.applications import vgg16

K.set_learning_phase(0)
input_tensor = Input((224, 224, 3))
model = vgg16.VGG16(input_tensor=input_tensor, include_top=False)
x = model.output
x = Flatten()(x)
x = Dense(1024, activation='relu')(x)
x = Dropout(0.5)(x)
x = Dense(512,activation='relu')(x)
x = Dropout(0.5)(x)
x = Dense(256,activation='relu')(x)
x = Dropout(0.5)(x)
x = Dense(1)(x)
model2 = Model(model.input, x)
model2.predict(np.zeros((1, 224, 224, 3)))
```

Produces:
```
$ python test-keras.py 
Using TensorFlow backend.
Traceback (most recent call last):
  File "test-keras.py", line 15, in <module>
    x = Dense(512,activation='relu')(x)
  File "/Users/todd/miniconda2/envs/landmarker/lib/python2.7/site-packages/keras/engine/topology.py", line 529, in __call__
    self.assert_input_compatibility(x)
  File "/Users/todd/miniconda2/envs/landmarker/lib/python2.7/site-packages/keras/engine/topology.py", line 457, in assert_input_compatibility
    if K.ndim(x) < ndim:
  File "/Users/todd/miniconda2/envs/landmarker/lib/python2.7/site-packages/keras/backend/tensorflow_backend.py", line 399, in ndim
    dims = x.get_shape()._dims
AttributeError: 'function' object has no attribute 'get_shape'
```

**Comment by kyle-dorman** [comment]
Hullo, 

I ran into a the same problem with Dropout layers when setting the session.

For example:
```
config = tf.ConfigProto(device_count = {'GPU': 0})
    with tf.Session(config=config) as sess:
        K.set_session(sess)
```

**Comment by IstrateRoxana** [comment]
I have the same problem when running this simple model:
```inp = layers.Input(shape=(17,))
d1 = layers.Dense(128, activation='relu')(inp)
d1_drop = layers.Dropout(0.5)(d1)
d2 = layers.Dense(512, activation='relu')(d1_drop)
d2_drop = layers.Dropout(0.5)(d2)
predictions = layers.Dense(17, activation='sigmoid')(d2_drop)
model = Model(inputs=inp, outputs=predictions)

model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
```

The error: 

```Traceback (most recent call last):
  File "train.py", line 199, in <module>
    d2 = layers.Dense(512, activation='relu')(d1_drop)
  File "/hpc_modules/at9.0-based/python-packages-3.4/lib64/python3.4/site-packages/keras/engine/topology.py", line 541, in __call__
    self.assert_input_compatibility(inputs)
  File "/hpc_modules/at9.0-based/python-packages-3.4/lib64/python3.4/site-packages/keras/engine/topology.py", line 450, in assert_input_compatibility
    ndim = K.ndim(x)
  File "/hpc_modules/at9.0-based/python-packages-3.4/lib64/python3.4/site-packages/keras/backend/tensorflow_backend.py", line 479, in ndim
    dims = x.get_shape()._dims
AttributeError: 'Dropout' object has no attribute 'get_shape'```

**Comment by fchollet** [comment]
`AttributeError: 'Dropout' object has no attribute 'get_shape'`

This error means that you passed a `Dropout` layer instance to a layer, instead of passing a tensor. I think we need a better error message in this case.

**Comment by kyle-dorman** [comment]
Hi I added a new issue to track adding a better error message and included small proposed solution here #6589
